tachyon/crypto/commitments/fri/two_adic_fri_verifier.h

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2022 The Plonky3 Authors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.plonky3 and the LICENCE-APACHE.plonky3
+// file.
+
+#ifndef TACHYON_CRYPTO_COMMITMENTS_FRI_TWO_ADIC_FRI_VERIFIER_H_
+#define TACHYON_CRYPTO_COMMITMENTS_FRI_TWO_ADIC_FRI_VERIFIER_H_
+
+#include <tuple>
+#include <vector>
+
+#include "tachyon/base/bits.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_config.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_proof.h"
+#include "tachyon/math/geometry/dimensions.h"
+#include "tachyon/zk/air/plonky3/challenger/challenger.h"
+
+namespace tachyon::crypto {
+
+template <typename F, typename MMCS>
+struct CommitStep {
+  using Commitment = typename MMCS::Commitment;
+  F beta;
+  Commitment commit;
+  CommitPhaseProofStep<F, MMCS> opening;
+};
+
+template <typename F, typename MMCS>
+F VerifyQuery(size_t index, size_t log_max_height,
+              const TwoAdicFriConfig<MMCS>& config,
+              const std::vector<CommitStep<F, MMCS>>& steps,
+              const std::vector<std::tuple<size_t, F>>& reduced_openings) {
+  F folded_eval = F::Zero();
+  auto ro_iter = reduced_openings.begin();
+  auto ro_end = reduced_openings.end();
+
+  for (size_t step_idx = 0; step_idx < steps.size(); ++step_idx) {
+    size_t log_folded_height = log_max_height - step_idx - 1;
+    if (ro_iter != ro_end && std::get<0>(*ro_iter) == log_folded_height + 1) {
+      folded_eval += std::get<1>(*ro_iter);
+      ++ro_iter;
+    }
+    size_t index_sibling = index ^ 1;
+    size_t index_pair = index >> 1;
+    std::vector<F> evals(2, folded_eval);
```

Comment:
Please use `std::array<F, 2>`

Reply:
This requires `(Do)VerifyingOpeningProof` for `field_merkle_tree_mmcs.h` and `extension_field_merkle_tree_mmcs.h` and `mixed_matrix_commitment_scheme.h` to take in `absl::span<absl::span<>>` as a parameter instead of `absl::span<std::vector<>>`. `mmcs_.VerifyOpeningProof(` in `two_adic_fri_pcs.h` then uses `opened_values` which is a `std::vector<std::vector<>>`, which means every inner vector would have to be typed separately to `absl::span` in order for `opened_values` to be able to be inputted into `VerifyOpeningProof(`. Do you think this change should still be done?